File: mvvm/src/main/java/com/lookballs/mvvm/action/ResourcesAction.java
package com.lookballs.mvvm.action;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.annotation.DimenRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;

/**
 * Context 意图处理（扩展非 Context 类的方法，请不要用 Context 子类实现此接口）
 */
public interface ResourcesAction {

    Context getContext();

    default Resources getResources() {
        return getContext().getResources();
    }

    default String getString(@StringRes int id) {
        return getResources().getString(id);
    }

    default String getString(@StringRes int id, Object... formatArgs) {
        return getResources().getString(id, formatArgs);
    }

    default float getDimension(@DimenRes int id) {
        return getResources().getDimension(id);
    }

    default int getDimensionPixelSize(@DimenRes int id) {
        return getResources().getDimensionPixelSize(id);
    }

    default Drawable getDrawable(@DrawableRes int id) {
        return ContextCompat.getDrawable(getContext(), id);
    }

    @ColorInt
    default int getColor(@ColorRes int id) {
        return ContextCompat.getColor(getContext(), id);
    }

    default <S> S getSystemService(@NonNull Class<S> serviceClass) {
        return ContextCompat.getSystemService(getContext(), serviceClass);
    }
}

File: mvvm-kt/src/main/java/com/lookballs/mvvm/action/ResourcesAction.kt
package com.lookballs.mvvm.action

import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.graphics.drawable.Drawable
import androidx.annotation.*
import androidx.core.content.ContextCompat

/**
 * Context 意图处理（扩展非 Context 类的方法，请不要用 Context 子类实现此接口）
 */
interface ResourcesAction {
    val `$context`: Context

    val `$resources`: Resources
        get() = `$context`.resources

    fun `$getString`(@StringRes id: Int): String? {
        return `$context`.getString(id)
    }

    fun `$getString`(@StringRes id: Int, vararg formatArgs: Any?): String? {
        return `$resources`.getString(id, *formatArgs)
    }

    fun `$getDimension`(@DimenRes id: Int): Float {
        return `$resources`.getDimension(id)
    }

    fun `$getDimensionPixelSize`(@DimenRes id: Int): Int {
        return `$resources`.getDimensionPixelSize(id)
    }

    fun `$getDrawable`(@DrawableRes id: Int): Drawable? {
        return ContextCompat.getDrawable(`$context`, id)
    }

    @ColorInt
    fun `$getColor`(@ColorRes id: Int): Int {
        return ContextCompat.getColor(`$context`, id)
    }

    fun `$getColorStateList`(@ColorRes id: Int): ColorStateList? {
        return ContextCompat.getColorStateList(`$context`, id)
    }

    fun <S> `$getSystemService`(serviceClass: Class<S>): S? {
        return ContextCompat.getSystemService(`$context`, serviceClass)
    }
}